Remove `ssr.paths` configuration and replace with `ssr.excludePathPatterns` which excludes specific paths from SSR

FrancescoMolinaro opened the pull request:
## References
Fixes #4132
Fixes #4042 

## Description

Adapted filtering of SSR pages blocking the rendering on specific pages instead of allowing it only on some pages.

## Instructions for Reviewers

Navigating on pages 500, 404 or 403, or any invalid URL I should see the correct error status in the request.
In general the response status should be always correct for pages rendered in SSR.

List of changes in this PR:

Removed `ssr.paths` configuration and replaced with `ssr.excludePathPatterns`.
Add patterns to exclude community and collection browse, global browse, global search, community list, and statistics and various administrative tools.

For example:
```
ssr:
    excludePathPatterns:
    # Excludes community browse pages/tabs
    - pattern: "^/communities/[a-f0-9-]{36}/browse(/.*)?$",
      flag: "i"
    # Excludes collection browse pages/tabs
    - pattern: "^/collections/[a-f0-9-]{36}/browse(/.*)?$"
      flag: "i"
    # Excludes site-wide browse-by pages
    - pattern: "^/browse/"
    # Excludes site-wide search
    - pattern: "^/search$"
    # Exclude community-list page
    - pattern: "^/community-list$"
    # Exclude various tools that are under the /admin path
    - pattern: "^/admin/"
    # Exclude processes
    - pattern: "^/processes/?"
    # Exclude QA Notifications
    - pattern: "^/notifications/"
    # Exclude all statistics pages (site-wide and for individual objects)
    - pattern: "^/statistics/?"
    # Exclude access-control tools
    - pattern: "^/access-control/"
    # Exclude health page
    - pattern: "^/health$"
```

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You do not need to complete this checklist prior creating your PR (draft PRs are always welcome).
However, reviewers may request that you complete any actions in this list if you have not done so. If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [ ] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [ ] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [ ] My PR **passes [ESLint](https://eslint.org/)** validation using `npm run lint`
- [ ] My PR **doesn't introduce circular dependencies** (verified via `npm run check-circ-deps`)
- [ ] My PR **includes [TypeDoc](https://typedoc.org/) comments** for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [ ] My PR **passes all specs/tests and includes new/updated specs or tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] My PR **aligns with [Accessibility guidelines](https://wiki.lyrasis.org/display/DSDOC8x/Accessibility)** if it makes changes to the user interface.
- [ ] My PR **uses i18n (internationalization) keys** instead of hardcoded English text, to allow for translations.
- [ ] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [ ] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [ ] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.
- [ ] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

tdonohue commented on server.ts:
```diff
@@ -406,10 +407,26 @@ function cacheCheck(req, res, next) {
+ * @param res
+ */
+function setRequestErrorStatus(req, res) {
+  if (req.url === '/500') {
```
@FrancescoMolinaro : Minor feedback, but I wonder if it'd be better to import these error paths from `app-routing-paths` like:
```
import { 
  getForbiddenRoute,
  getPageNotFoundRoute,
  getPageInternalServerErrorRoute
} from './app-routing-paths';
```

That way we avoid hardcoding these paths in two places, and allow for the paths to change if they are ever modified in `app-routing-paths.ts`.

FrancescoMolinaro replied:
Hi @tdonohue , thanks for the feedback! Your point is totally fair, I have therefore replaced the hardcoded path with the method you suggested, so to improve maintenability.

I will look forward to any other feedback and thanks again.

tdonohue reviewed (changes requested):
@FrancescoMolinaro : The code looks good overall now, but my tests are not all successful.  Here's what I'm seeing:

1. Running this PR in production mode (`npm run build:prod; npm run serve:ssr`)
2. Test http://localhost:4000/500 -> Returns 500 error code (SUCCESS)
3. Test http://localhost:4000/403 -> Returns 403 error code (SUCCESS)
4. Test http://localhost:4000/404 -> Returns 404 error code (SUCCESS)
5. Test http://localhost:4000/invalid-url -> Shows the `/404` page, but **returns 200 OK** (FAILURE)

So, this works for the pages that it's checking for.  But, it doesn't seem to work in scenarios where a different path results in a 404 error. It also therefore might not work if a different path resulted in a 500 or 403 error.

Overall, this PR provides **better behavior** than the current `main`.  But, it doesn't seem to fully fix the problem of returning a proper error code when an error occurs from the backend.  Instead, it just makes sure that if you access these specific error pages *directly in your browser* that you'll get a proper error code.

tdonohue commented:
Assigning @artlowel as a secondary reviewer since he offered to help take a look at this today.

AndreaBarbasso commented:
@tdonohue, @artlowel, the same situation (using a 200 error code for something that is not explicitly - eg. /404, /500 - a redirect/error) can be seen here: https://github.com/DSpace/dspace-angular/issues/4042.
I think @FrancescoMolinaro is on the right track, we might want to re-think the solution applied [here](https://github.com/DSpace/dspace-angular/pull/3682) and use a black-list approach instead of a white-list approach.

tdonohue commented:
@AndreaBarbasso and @FrancescoMolinaro :  If we can find a way to solve this by changing our settings to *exclude* specific paths from SSR, instead of *including* only specific paths in SSR, then I'd be OK with that approach.  That does seem like it may provide better behavior for unexpected errors / redirects.

If we want to go that route, I'd recommend renaming the existing `ssr.paths` setting in `config.*.yml` to make the new behavior clear.  Perhaps calling it something like `ssr.excludePaths` or similar.  So, something like this:
```
ssr:
  excludePaths: [ '/browse/', '/community-list', '/search/', '/statistics']
```
The challenge may be that we may need to also support regex values if we also want to exclude browse pages for communities & collections which appear under `/communities/[uuid]/browse/` and `/collections/[uuid]/browse` respectively.

Whatever solution is created, we'll have to make sure to also backport it to 7.6.x and 8.x.

FrancescoMolinaro commented:
Hi @tdonohue , thanks for the feedback!

I agree that a blacklist approach would make the solution more stable for error codes and redirects, so I have refactored the implementation following your suggestion.
I still think there will be some edge cases where we can't show a correct error code, e.g. redirect to login page if a user access a restricted collection or community, but I believe this change will anyway reduce considerably the number of these cases.

I will look forward to your feedback and thanks again!

tdonohue commented on server.ts:
```diff
@@ -627,6 +627,17 @@ function start() {
+ * @param path
+ * @param blacklist
+ */
+function isBlacklistedPathForSsr(url: string, path: string, blacklist: RegExp[]): boolean {
```
We should remove the term "blacklist" from our code, as it can have [negative connotations](https://www.splunk.com/en_us/blog/learn/blacklist-whitelist-inclusivity.html).

Please change the terms in your code to either say "exclusion list" or "block list" or similar.  Those terms are more descriptive as well.  So, this method could be named `isExcludedFromSsr()` and the last param could be called `exclusionList` or `excludeRegex` or similar.

FrancescoMolinaro commented:
Hi @tdonohue , thanks for the feedback and sorry for the poor naming choice, I appreciated a lot the article you shared.

I have refactored the solution so to have a single parameter based on regexes as you proposed, I agree that this give a better flexibility and power in handling the various paths compositiions.

I will look forward to your next feedback and thanks again.

tdonohue commented on config/config.example.yml:
```diff
@@ -23,10 +23,9 @@ ssr:
-  paths: [ '/home', '/items/', '/entities/', '/collections/', '/communities/', '/bitstream/', '/bitstreams/', '/handle/', '/reload/' ]
+  # Regexes to be run against the path of the page to check if SSR is allowed.
+  # If the path match any of the regexes it will be served directly in CSR.
+  excludePathRegexes: [ /^\/communities\/[0-9a-fA-F-]{36}(\/.*)?$/, /^\/collections\/[0-9a-fA-F-]{36}(\/.*)?$/]
```
Can we updated this to have the same list as your have in `environment.production.ts`?  I'd also recommend we add a basic explanation of these regexes. Something like:
```
# By default, excludes community and collection browse, global browse, global search, community list, and statistics.
```

tdonohue commented on server.ts:
```diff
@@ -221,7 +221,7 @@ export function app() {
  */
 function ngApp(req, res, next) {
-  if (environment.ssr.enabled && req.method === 'GET' && (req.path === '/' || environment.ssr.paths.some(pathPrefix => req.path.startsWith(pathPrefix)))) {
+    if (environment.ssr.enabled && req.method === 'GET' && (req.path === '/' || !isExcludedFromSsr(req.path, environment.ssr.excludePathRegexes))) {
```
There's an extra space before this `if` that seems unnecessary. I'm surprised that `lint` didn't catch it.

tdonohue reviewed (commented):
@FrancescoMolinaro : Thanks for the updates.  Overall, this now looks good to me.  I only have minor suggestions inline below.

I'd also recommend we **update** the Title & Description of this PR, as it's changed completely from its original purpose.  If possible, we may also want to remove the earlier commits in this PR (by rebasing the PR & squashing commits).  But, if you aren't able to easily do that, I can always merge squash this (reducing the entire PR to a single commit).

(UPDATE: I've changed the PR title. But, I haven't yet updated the description)

Overall, though, I'm +1.  I also tested this today and verified that this approach works.  I can verify that the excluded paths are *not* triggering SSR.  I can also verify that the error pages & invalid URLs are returning correct error codes again.

FrancescoMolinaro commented:
Hi @tdonohue , thanks again for the feedback.

I did the last changes you requested and I have updated the description and squashed all the commits so that we now have a single one.
The PR should be now in a cleaner state but please let me know if there is something else that needs improvement.

tdonohue commented on config/config.example.yml:
```diff
@@ -23,10 +23,17 @@ ssr:
+  # If the path match any of the regexes it will be served directly in CSR.
+  # By default, excludes community and collection browse, global browse, global search, community list, and statistics.
+  excludePathRegexes: [
+    /^\/communities\/[ 0-9a-f ]{ 8 }-[ 0-9a-f ]{ 4 }-[ 0-9a-f ]{ 4 }-[ 0-9a-f ]{ 4 }-[ 0-9a-f ]{ 12 }\/browse(\/.*)?$/i,
```
I'm unable to get this syntax to work in our YAML configuration file.  In fact, I've tried a variety of options, including this syntax, or structuring it like:
```
ssr:
  excludePathRegexes:
    - /^\/communities\/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\/browse(\/.*)?$/i
    - /^\/collections\/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\/browse(\/.*)?$/i
    - /^\/browse\//
    - /^\/search$/
    - /^\/community-list$/
    - /^\/statistics\//
```

However, even when I use that syntax, as soon as `isExcludedFromSsr()` is run, I always seen an error that says:
```
regex.test is not a function
```

It almost seems like the YAML config is being parsed as a String, and therefore cannot be run as a RegEx using `test(path)`.

Is there something else we need to do in order to override these RegEx values in our config file?  Or do I have the syntax wrong?

tdonohue reviewed (changes requested):
@FrancescoMolinaro : Apologies for not getting back to this PR for a while.  I re-tested it today, and it's mostly looking good & working well. However, I've run in a few more small bugs described inline below.  Basically, we need to make a small update to the "/statistics" regex, and I'm also having issues with overriding the default settings in my `config.prod.yml`.

Beyond that, I'm still basically a +1.  I just want to ensure we get these final issues fixed before merging.

atarix83 commented:
@FrancescoMolinaro 
I'd exclude also the following paths:
- /admin
- /processes
- /notifications
- /health

FrancescoMolinaro commented:
Hi @tdonohue , I checked and actually YAML doesn't support RegEx so they are treated as strings.
I have improved the solution so that the patterns can be directly RegEx or strings, I have tested your YAML configuration and seems to work.
In addition I added to the patterns also the ones suggested by @atarix83 .

tdonohue commented on config/config.example.yml:
```diff
@@ -23,10 +23,21 @@ ssr:
-  paths: [ '/home', '/items/', '/entities/', '/collections/', '/communities/', '/bitstream/', '/bitstreams/', '/handle/', '/reload/' ]
+  # Patterns to be run as regexes against the path of the page to check if SSR is allowed.
+  # If the path match any of the regexes it will be served directly in CSR.
+  # By default, excludes community and collection browse, global browse, global search, community list, and statistics.
```
Now that we've added more patterns, we should update this sentence to end with "community list, statistics and various administrative tools"

tdonohue reviewed (changes requested):
@FrancescoMolinaro : Thanks for the updates.  The basics appear to work, but I'm still **not able** to properly override settings using my `config.*.yml`. 

I'm testing in Production Mode (`npm run build:prod; npm run serve:ssr`), and I'm turning **off** Javascript in my browser and attempting to access various paths which should be excluded.  

Here's what I'm trying:

Add the following YAML settings:
```
ssr:
  excludePathPatterns:
    - /^\/community-list$/
```

Now, attempt to access http://localhost:4000/community-list and it **wrongly** is served via SSR.  

However, if I **comment out** those YAML settings and use the defaults, then http://localhost:4000/community-list is correctly blocked from SSR.

So, it appears overrides in YAML are not working properly still.  The default settings do work, but I cannot figure out how to override defaults.

I've also noticed a few of the new patterns appear to be incorrect. I believe these all need to be corrected
```
# There's no "/admin" path, but lots of paths like "/admin/[subpath]"
/^\/admin\//
# This needs to match both "/processes" and "/processes/[id]"
/^\/processes\/?/
# There's no "/notifications" path, but a few paths like "/notifications/[subpath]"
/^\/notifications\//
# I gave the wrong pattern for this one, as we also need to match "/statistics"
/^\/statistics\/?/
# For completion, we may also want to add this as other admin tools use "/access-control/[subpath]".
/^\/access-control\//
```
(Please double check my regex patterns are correct.  I *think* these are correct, but I also might have made a mistake.  I wasn't able to easily test these because the YAML overrides are not working.)

Finally, one note inline below.

philipprumpf commented on config/config.example.yml:
```diff
@@ -23,10 +23,21 @@ ssr:
+  # If the path match any of the regexes it will be served directly in CSR.
+  # By default, excludes community and collection browse, global browse, global search, community list, and statistics.
+  excludePathPatterns:
+    - /^\/communities\/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\/browse(\/.*)?$/i
```
My apologies if this is a local problem, but are we certain that using the extended regexp syntax here (with slashes and final flags) actually works the way it should? Yaml doesn't know about regular expressions (so treats them as unquoted strings), and the RegExp constructor does not expect the initial or final slashes and/or flags, as far as I know.

It's possible this is a local problem or one specific to how I run yarn, but `console.log` appears to confirm there's a problem with the example paths, when they're provided in a .yml

tdonohue commented:
@philipprumpf : Thanks for digging in here & figuring out why this PR is no longer working.  I see similar behavior.

@FrancescoMolinaro : Could you update this PR to no longer use the `/i` regex (I prefer @philipprumpf 's first option listed above), and see if you can solve the issues I've mentioned above about overriding the default values?  We will need to move quickly on this PR if we're going to get it merged for 9.0  (merger deadline is Friday)

FrancescoMolinaro commented:
Hi @tdonohue, @philipprumpf , thanks for the feedback and sorry for the delay, I wasn't working the past few days.

I have refactored the solution following the suggestions of @philipprumpf to prevent the slashes from the strings in the YAML and adapted a bit the logic so that we can pass the flag as an additional parameter, therefore adding a bit more flexibility in building the RegExp.

After running some tests it seem working to me, I tried accessing every path that should be covered but it would be great if you could double check them.

Regarding the issue with YAML override I believe it was connected with the fact that the patterns would have a different value when passed from the YAML, I tried the patterns you suggested and they seem correct and working when applied from the YAML.

I will look forward to your next feedback and thanks again.

tdonohue reviewed (approved):
👍 Thanks @FrancescoMolinaro !  This looks good to me now.  I've retested today, and all the prior issues are fixed.  The error codes are working properly now for the `/404`, `/403` and `/500` error pages (and any invalid URLs). I've also verified that overriding the default settings in your `config.*.yml` now works properly.

tdonohue commented:
@FrancescoMolinaro : Could you manually backport this to `dspace-7_x` please?  It was only able to be automatically backported to 8.x.

FrancescoMolinaro commented:
Hi @tdonohue , I have opened a PR for the backporting towards dspace-7_x here https://github.com/DSpace/dspace-angular/pull/4332.

toniprieto commented on config/config.example.yml:
```diff
@@ -23,10 +23,24 @@ ssr:
+  # If the path match any of the regexes it will be served directly in CSR.
+  # By default, excludes community and collection browse, global browse, global search, community list, statistics and various administrative tools.
+  excludePathPatterns:
+    - pattern: "^/communities/[a-f0-9-]{36}/browse(/.*)?$",
```
I was testing this functionality and it wasn't reading the configuration correctly. I had to remove the trailing comma from this line — I guess it's a copy & paste error and it shouldn't be there?

tdonohue replied:
@toniprieto : Yes, you are correct. That appears to be a copy & paste error.  The comma should not be there. Would you have time to create a small PR to fix this?  If not, I can do so.

toniprieto replied:
Yes, I'll do it now.

tdonohue replied:
Created a PR at https://github.com/DSpace/dspace-angular/pull/4364  Will quickly fix in 9.0 prior to release
